File: ruby/triangle/triangle.rb
class Triangle
  def initialize(sides)
    @sides = sides.sort
  end

  def equilateral?
    valid? && unique_sides == 1
  end

  def isosceles?
    valid? && unique_sides <= 2
  end

  def scalene?
    valid? && unique_sides == 3
  end

  def valid?
    a, b, c = @sides

    return @sides.all? { |s| s > 0 } && a + b >= c
  end

  def unique_sides
    @sides.uniq.size
  end
end

File: cpp/triangle/triangle.cpp
#include "triangle.h"
#include <set>
#include <array>
#include <algorithm>
#include <stdexcept>

using namespace std;

triangle::flavor triangle::kind(float a, float b, float c) {
    array<float, 3> sides = {a, b, c};

    sort(sides.begin(), sides.end());

    if (sides[0] <= 0)
        throw domain_error("Each side must have a positive length");

    if (sides[0] + sides[1] < sides[2])
        throw domain_error("Any two sides of the triangle must be greater than or equal to the remaining side");

    set<float> uniq {sides.begin(), sides.end()};

    switch (uniq.size()) {
    case 1:
        return triangle::flavor::equilateral;
    case 2:
        return triangle::flavor::isosceles;
    case 3:
        return triangle::flavor::scalene;
    default:
        return triangle::flavor::invalid;
    }
}
